File: food-for-thought/tesseract/determine-menu.py
from PIL import Image
import pytesseract
import re


def extract_text_from_image(image_path) -> str:
    img = Image.open(image_path)
    extracted_text = pytesseract.image_to_string(img)
    return extracted_text

"""
Determines if the image is a menu or not.

The function checks the price + the headers to check if item is a menu.

Input: extracted_text from the extract_text_from_image function
Output: True if the image is menu, false if it isn't
"""
def is_menu(extracted_text: str) -> bool:
    common_keywords = ['appetizers', 'appetizer', 'main course', 'desserts', 'starters',
                'salads', 'drinks', 'entrees', 'sides', 'beverages', 'soups', 'mains',
                'lunch', 'menu']
    
    # common price pattern
    price_pattern = r'\$\d+(\.\d{2})?'
    text_lower = extracted_text.lower()

    # another method of implementation but is potentially less accurate
    # if(re.search(common_keywords, text_lower) and re.search(price_pattern, text_lower)):
    #     return True
    
    common_keyword_count = sum([text_lower.count(keyword.lower()) for keyword in common_keywords])
    price_count = len(re.findall(price_pattern, text_lower))
    
    # is a menu if more than 1 keyword is found & has more than 3 prices on the img
    if common_keyword_count >= 1 or price_count >= 3:
        return True
    
    return False

File: food-for-thought/tesseract/determine-menu.js
const Tesseract = require('tesseract.js');
const fs = require('fs');

async function extractTextFromImage(imagePath) {
    try {
        const result = await Tesseract.recognize(imagePath, 'eng');
        const extractedText = result.data.text;
        console.log(extractedText); // Print the extracted text
        return extractedText;
    } catch (error) {
        console.error("Error extracting text:", error);
        return '';
    }
}

// Function to determine if the text represents a menu
function isMenu(extractedText) {
    const commonKeywords = ['appetizers', 'appetizer', 'main course', 'desserts', 'starters',
                            'salads', 'drinks', 'entrees', 'sides', 'beverages', 'soups', 
                            'mains', 'lunch', 'menu', 'vegetables'];
    
    // Regular expression for price pattern
    const pricePattern = /\$\d+(\.\d{2})?/g;
    const textLower = extractedText.toLowerCase();
    const commonKeywordCount = commonKeywords.reduce((count, keyword) => {
        return count + (textLower.includes(keyword) ? 1 : 0);
    }, 0);

    const priceCount = (textLower.match(pricePattern) || []).length;

    if (commonKeywordCount >= 1 && priceCount >= 3) {
        console.log("Item is a menu");
        return true;
    }

    console.log("Item is not a menu");
    return false;
}

// FOR TESTING
const imagePath = "";
extractTextFromImage(imagePath).then((extractedText) => {
    console.log("Extracted Text:");
    console.log(extractedText);
    isMenu(extractedText);
});
